pkg/common/telemetry/metrics.go

```diff
@@ -100,62 +100,81 @@ func (m *MetricsImpl) ListenAndServe(ctx context.Context) error {
 
 func (m *MetricsImpl) SetGauge(key []string, val float32) {
 	for _, s := range m.metricsSinks {
-		s.SetGauge(key, val)
+		keyCopy := copyKey(key)
+		s.SetGauge(keyCopy, val)
 	}
 }
 
 // SetGaugeWithLabels delegates to embedded metrics, sanitizing labels
 func (m *MetricsImpl) SetGaugeWithLabels(key []string, val float32, labels []Label) {
 	sanitizedLabels := SanitizeLabels(labels)
 	for _, s := range m.metricsSinks {
-		s.SetGaugeWithLabels(key, val, sanitizedLabels)
+		keyCopy := copyKey(key)
+		s.SetGaugeWithLabels(keyCopy, val, sanitizedLabels)
 	}
 }
 
 func (m *MetricsImpl) EmitKey(key []string, val float32) {
 	for _, s := range m.metricsSinks {
-		s.EmitKey(key, val)
+		keyCopy := copyKey(key)
+		s.EmitKey(keyCopy, val)
 	}
 }
 
 func (m *MetricsImpl) IncrCounter(key []string, val float32) {
 	for _, s := range m.metricsSinks {
-		s.IncrCounter(key, val)
+		keyCopy := copyKey(key)
+		s.IncrCounter(keyCopy, val)
 	}
 }
 
 // IncrCounterWithLabels delegates to embedded metrics, sanitizing labels
 func (m *MetricsImpl) IncrCounterWithLabels(key []string, val float32, labels []Label) {
 	sanitizedLabels := SanitizeLabels(labels)
 	for _, s := range m.metricsSinks {
-		s.IncrCounterWithLabels(key, val, sanitizedLabels)
+		keyCopy := copyKey(key)
+		s.IncrCounterWithLabels(keyCopy, val, sanitizedLabels)
 	}
 }
 
 func (m *MetricsImpl) AddSample(key []string, val float32) {
 	for _, s := range m.metricsSinks {
-		s.AddSample(key, val)
+		keyCopy := copyKey(key)
+		s.AddSample(keyCopy, val)
 	}
 }
 
 // AddSampleWithLabels delegates to embedded metrics, sanitizing labels
 func (m *MetricsImpl) AddSampleWithLabels(key []string, val float32, labels []Label) {
 	sanitizedLabels := SanitizeLabels(labels)
 	for _, s := range m.metricsSinks {
-		s.AddSampleWithLabels(key, val, sanitizedLabels)
+		keyCopy := copyKey(key)
+		s.AddSampleWithLabels(keyCopy, val, sanitizedLabels)
 	}
 }
 
 func (m *MetricsImpl) MeasureSince(key []string, start time.Time) {
 	for _, s := range m.metricsSinks {
-		s.MeasureSince(key, start)
+		keyCopy := copyKey(key)
+		s.MeasureSince(keyCopy, start)
 	}
 }
 
 // MeasureSinceWithLabels delegates to embedded metrics, sanitizing labels
 func (m *MetricsImpl) MeasureSinceWithLabels(key []string, start time.Time, labels []Label) {
 	sanitizedLabels := SanitizeLabels(labels)
 	for _, s := range m.metricsSinks {
-		s.MeasureSinceWithLabels(key, start, sanitizedLabels)
+		keyCopy := copyKey(key)
+		s.MeasureSinceWithLabels(keyCopy, start, sanitizedLabels)
 	}
 }
+
+// TODO: [rturner3] Remove this if/when go-metrics is fixed to not modify its input "key" slice
+func copyKey(key []string) []string {
+	var newKey []string
```

Comment:
The following is a much more efficient way to copy the slice:
```
func copyKey(key []string) string {
   return append([]string(nil), key...)
}
```

You can get similar perf characteristics by preallocating the `newKey` slice with enough capacity to hold `key` but the above is much more terse.